File: sdk/rust/src/schema.rs
use serde_json::{json, Value};
use std::collections::{HashMap, HashSet};
use std::sync::{Arc, Mutex};

use crate::types::{ParseResult, PathSegment, UnknownKeyMode, ValidationError, ValidationIssue};

/// Context passed through during parsing, carrying definitions for ref resolution.
#[derive(Debug, Clone)]
pub struct ParseContext {
    pub definitions: HashMap<String, Box<dyn Schema>>,
    pub inherited_unknown_keys: Option<UnknownKeyMode>,
    active_refs: Arc<Mutex<HashSet<String>>>,
}

impl ParseContext {
    pub fn new() -> Self {
        ParseContext {
            definitions: HashMap::new(),
            inherited_unknown_keys: None,
            active_refs: Arc::new(Mutex::new(HashSet::new())),
        }
    }

    pub fn with_definitions(definitions: HashMap<String, Box<dyn Schema>>) -> Self {
        ParseContext {
            definitions,
            inherited_unknown_keys: None,
            active_refs: Arc::new(Mutex::new(HashSet::new())),
        }
    }

    pub fn enter_ref(&self, key: String) -> bool {
        self.active_refs.lock().unwrap().insert(key)
    }

    pub fn exit_ref(&self, key: &str) {
        self.active_refs.lock().unwrap().remove(key);
    }
}

impl Default for ParseContext {
    fn default() -> Self {
        Self::new()
    }
}

/// The core Schema trait. All schema types implement this.
pub trait Schema: SchemaClone + std::fmt::Debug + Send + Sync {
    /// The kind string for this schema (e.g., "string", "int", "object").
    fn kind(&self) -> &str;

    /// Validate and parse the input value. Returns the (possibly transformed) value or issues.
    fn parse_value(
        &self,
        input: &Value,
        path: &[PathSegment],
        ctx: &ParseContext,
    ) -> Result<Value, Vec<ValidationIssue>>;

    /// Throwing parse: returns Ok(Value) or Err(ValidationError).
    fn parse(&self, input: &Value) -> Result<Value, ValidationError> {
        self.parse_with_context(input, &ParseContext::new())
    }

    /// Throwing parse with context (for refs).
    fn parse_with_context(
        &self,
        input: &Value,
        ctx: &ParseContext,
    ) -> Result<Value, ValidationError> {
        match self.parse_value(input, &[], ctx) {
            Ok(v) => Ok(v),
            Err(issues) => Err(ValidationError { issues }),
        }
    }

    /// Non-throwing parse: returns ParseResult with success/failure info.
    fn safe_parse(&self, input: &Value) -> ParseResult {
        self.safe_parse_with_context(input, &ParseContext::new())
    }

    /// Non-throwing parse with context.
    fn safe_parse_with_context(&self, input: &Value, ctx: &ParseContext) -> ParseResult {
        match self.parse_value(input, &[], ctx) {
            Ok(v) => ParseResult::ok(v),
            Err(issues) => ParseResult::err(issues),
        }
    }

    /// Export this schema to a JSON node representation.
    fn export_node(&self) -> Value;

    /// Schema-level default used by wrappers that carry their own default.
    fn default_value(&self) -> Option<&Value> {
        None
    }

    /// Whether this schema has non-portable features (custom validators).
    fn has_custom_validators(&self) -> bool {
        false
    }
}

/// Trait for cloning boxed schemas.
pub trait SchemaClone {
    fn clone_box(&self) -> Box<dyn Schema>;
}

impl<T: Schema + Clone + 'static> SchemaClone for T {
    fn clone_box(&self) -> Box<dyn Schema> {
        Box::new(self.clone())
    }
}

impl Clone for Box<dyn Schema> {
    fn clone(&self) -> Box<dyn Schema> {
        self.clone_box()
    }
}

/// Reserved metadata keys that must use describe() instead of metadata().
pub fn reserved_metadata_keys() -> HashSet<&'static str> {
    let mut keys = HashSet::new();
    keys.insert("title");
    keys.insert("description");
    keys.insert("deprecated");
    keys.insert("deprecatedMessage");
    keys.insert("notStable");
    keys.insert("since");
    keys.insert("sensitive");
    keys.insert("readonly");
    keys.insert("writeonly");
    keys.insert("examples");
    keys
}

/// Options for describe().
#[derive(Debug, Clone, Default)]
pub struct DescribeOpts {
    pub title: Option<String>,
    pub deprecated: Option<bool>,
    pub deprecated_message: Option<String>,
    pub not_stable: Option<bool>,
    pub since: Option<String>,
    pub sensitive: Option<bool>,
    pub readonly: Option<bool>,
    pub writeonly: Option<bool>,
    pub examples: Option<Vec<Value>>,
}

/// Validate and build metadata from describe() parameters.
pub fn build_describe_metadata(description: &str, opts: Option<&DescribeOpts>) -> Value {
    let mut meta = serde_json::Map::new();
    meta.insert("description".to_string(), json!(description));

    if let Some(opts) = opts {
        if let Some(title) = &opts.title {
            meta.insert("title".to_string(), json!(title));
        }
        if let Some(deprecated) = opts.deprecated {
            meta.insert("deprecated".to_string(), json!(deprecated));
        }
        if let Some(msg) = &opts.deprecated_message {
            if opts.deprecated != Some(true) {
                panic!("describe(): deprecatedMessage requires deprecated to be true");
            }
            meta.insert("deprecatedMessage".to_string(), json!(msg));
        }
        if let Some(not_stable) = opts.not_stable {
            meta.insert("notStable".to_string(), json!(not_stable));
        }
        if let Some(since) = &opts.since {
            meta.insert("since".to_string(), json!(since));
        }
        if let Some(sensitive) = opts.sensitive {
            meta.insert("sensitive".to_string(), json!(sensitive));
        }
        if let Some(readonly) = opts.readonly {
            meta.insert("readonly".to_string(), json!(readonly));
        }
        if let Some(writeonly) = opts.writeonly {
            meta.insert("writeonly".to_string(), json!(writeonly));
        }
        if opts.readonly == Some(true) && opts.writeonly == Some(true) {
            panic!("describe(): readonly and writeonly cannot both be true");
        }
        if let Some(examples) = &opts.examples {
            meta.insert("examples".to_string(), json!(examples));
        }
    }

    Value::Object(meta)
}

/// Validate metadata keys (for metadata() call - no reserved keys allowed).
pub fn validate_metadata_keys(meta: &Value) {
    let reserved = reserved_metadata_keys();
    if let Value::Object(map) = meta {
        for key in map.keys() {
            if reserved.contains(key.as_str()) {
                panic!(
                    "metadata(): \"{}\" is a reserved key. Use describe() instead.",
                    key
                );
            }
        }
    }
}

/// Merge metadata: shallow merge src into dst.
pub fn merge_metadata(dst: &mut Value, src: &Value) {
    if let (Value::Object(d), Value::Object(s)) = (dst, src) {
        for (k, v) in s {
            d.insert(k.clone(), v.clone());
        }
    }
}

/// Helper to add metadata to an export node.
pub fn add_metadata_to_node(node: &mut Value, metadata: &Option<Value>) {
    if let Some(meta) = metadata {
        if let Value::Object(map) = meta {
            if !map.is_empty() {
                if let Value::Object(obj) = node {
                    obj.insert("metadata".to_string(), meta.clone());
                }
            }
        }
    }
}

File: sdk/php/src/Schema.php
<?php

declare(strict_types=1);

namespace AnyVali;

use AnyVali\Interchange\Exporter;
use AnyVali\Parse\Coercion;

/**
 * @template TOutput
 */
abstract class Schema
{
    private const RESERVED_METADATA_KEYS = [
        'title', 'description', 'deprecated', 'deprecatedMessage',
        'notStable', 'since', 'sensitive', 'readonly', 'writeonly', 'examples',
    ];

    protected mixed $defaultValue = null;
    protected bool $hasDefault = false;
    /** @var string|string[]|null */
    protected string|array|null $coerce = null;
    /** @var callable[] */
    protected array $customValidators = [];
    /** @var array<string, mixed> */
    protected array $metadata = [];

    abstract public function getKind(): string;

    /**
     * Validate and return parsed value or issues.
     * @param ValidationContext $ctx
     * @return ParseResult
     */
    abstract protected function validateValue(mixed $value, ValidationContext $ctx): ParseResult;

    /**
     * Export this schema node to array form.
     * @return array<string, mixed>
     */
    abstract public function exportNode(): array;

    /**
     * Throwing parse.
     *
     * @return TOutput
     */
    public function parse(mixed $input): mixed
    {
        $result = $this->safeParse($input);
        if (!$result->success) {
            throw new ValidationError($result->issues);
        }
        return $result->value;
    }

    /**
     * Non-throwing parse.
     *
     * @return ParseResult<TOutput>
     */
    public function safeParse(mixed $input, ?ValidationContext $ctx = null): ParseResult
    {
        $ctx ??= new ValidationContext();
        $value = $input;

        // Step 1: Coercion (only if value is present)
        if ($this->coerce !== null) {
            $coercions = is_array($this->coerce) ? $this->coerce : [$this->coerce];
            foreach ($coercions as $c) {
                [$value, $issue] = Coercion::apply($c, $value, $this->getKind(), $ctx->path);
                if ($issue !== null) {
                    return ParseResult::fail([$issue]);
                }
            }
        }

        // Step 2: Validate
        $result = $this->validateValue($value, $ctx);

        // Step 3: Custom validators (local-only)
        if ($result->success && !empty($this->customValidators)) {
            $issues = [];
            foreach ($this->customValidators as $validator) {
                $validatorResult = $validator($result->value, $ctx);
                if ($validatorResult instanceof ValidationIssue) {
                    $issues[] = $validatorResult;
                }
            }
            if (!empty($issues)) {
                return ParseResult::fail($issues);
            }
        }

        return $result;
    }

    /**
     * Export as AnyValiDocument.
     */
    public function export(ExportMode $mode = ExportMode::Portable): AnyValiDocument
    {
        if ($mode === ExportMode::Portable && !empty($this->customValidators)) {
            throw new \RuntimeException(
                'Cannot export schema with custom validators in portable mode'
            );
        }
        return Exporter::export($this, $mode);
    }

    /**
     * Set default value. Returns cloned instance.
     */
    public function default(mixed $value): static
    {
        $clone = clone $this;
        $clone->defaultValue = $value;
        $clone->hasDefault = true;
        return $clone;
    }

    public function hasDefaultValue(): bool
    {
        return $this->hasDefault;
    }

    public function getDefaultValue(): mixed
    {
        return $this->defaultValue;
    }

    /**
     * Set coercion. Returns cloned instance.
     * @param string|string[] $coerce
     */
    public function coerce(string|array $coerce): static
    {
        $clone = clone $this;
        $clone->coerce = $coerce;
        return $clone;
    }

    public function getCoerce(): string|array|null
    {
        return $this->coerce;
    }

    /**
     * Add documentation metadata. Returns cloned instance.
     * @param array{title?: string, deprecated?: bool, deprecatedMessage?: string, notStable?: bool, since?: string, sensitive?: bool, readonly?: bool, writeonly?: bool, examples?: array} $opts
     */
    public function describe(string $description, array $opts = []): static
    {
        $clone = clone $this;
        $clone->metadata = array_merge($clone->metadata, ['description' => $description]);

        if (isset($opts['title'])) {
            if (!is_string($opts['title'])) {
                throw new \InvalidArgumentException('describe(): title must be a string');
            }
            $clone->metadata['title'] = $opts['title'];
        }
        if (isset($opts['deprecated'])) {
            if (!is_bool($opts['deprecated'])) {
                throw new \InvalidArgumentException('describe(): deprecated must be a boolean');
            }
            $clone->metadata['deprecated'] = $opts['deprecated'];
        }
        if (isset($opts['deprecatedMessage'])) {
            if (!is_string($opts['deprecatedMessage'])) {
                throw new \InvalidArgumentException('describe(): deprecatedMessage must be a string');
            }
            if (empty($opts['deprecated'])) {
                throw new \InvalidArgumentException('describe(): deprecatedMessage requires deprecated to be true');
            }
            $clone->metadata['deprecatedMessage'] = $opts['deprecatedMessage'];
        }
        if (isset($opts['notStable'])) {
            if (!is_bool($opts['notStable'])) {
                throw new \InvalidArgumentException('describe(): notStable must be a boolean');
            }
            $clone->metadata['notStable'] = $opts['notStable'];
        }
        if (isset($opts['since'])) {
            if (!is_string($opts['since'])) {
                throw new \InvalidArgumentException('describe(): since must be a string');
            }
            $clone->metadata['since'] = $opts['since'];
        }
        if (isset($opts['sensitive'])) {
            if (!is_bool($opts['sensitive'])) {
                throw new \InvalidArgumentException('describe(): sensitive must be a boolean');
            }
            $clone->metadata['sensitive'] = $opts['sensitive'];
        }
        if (isset($opts['readonly'])) {
            if (!is_bool($opts['readonly'])) {
                throw new \InvalidArgumentException('describe(): readonly must be a boolean');
            }
            $clone->metadata['readonly'] = $opts['readonly'];
        }
        if (isset($opts['writeonly'])) {
            if (!is_bool($opts['writeonly'])) {
                throw new \InvalidArgumentException('describe(): writeonly must be a boolean');
            }
            $clone->metadata['writeonly'] = $opts['writeonly'];
        }
        if (!empty($opts['readonly']) && !empty($opts['writeonly'])) {
            throw new \InvalidArgumentException('describe(): readonly and writeonly cannot both be true');
        }
        if (isset($opts['examples'])) {
            if (!is_array($opts['examples'])) {
                throw new \InvalidArgumentException('describe(): examples must be an array');
            }
            $clone->metadata['examples'] = $opts['examples'];
        }

        return $clone;
    }

    /**
     * Attach arbitrary metadata. Reserved keys must use describe().
     * @param array<string, mixed> $meta
     */
    public function metadata(array $meta, bool $replace = false): static
    {
        foreach (array_keys($meta) as $key) {
            if (in_array($key, self::RESERVED_METADATA_KEYS, true)) {
                throw new \InvalidArgumentException(
                    "metadata(): \"{$key}\" is a reserved key. Use describe() instead."
                );
            }
        }

        $clone = clone $this;
        if ($replace) {
            $preserved = [];
            foreach ($clone->metadata as $k => $v) {
                if (in_array($k, self::RESERVED_METADATA_KEYS, true)) {
                    $preserved[$k] = $v;
                }
            }
            $clone->metadata = array_merge($preserved, $meta);
        } else {
            $clone->metadata = array_merge($clone->metadata, $meta);
        }
        return $clone;
    }

    /**
     * @return array<string, mixed>
     */
    public function getMetadata(): array
    {
        return $this->metadata;
    }

    /**
     * Add a custom validator (local-only, non-portable).
     */
    public function refine(callable $validator): static
    {
        $clone = clone $this;
        $clone->customValidators[] = $validator;
        return $clone;
    }

    public function hasCustomValidators(): bool
    {
        return !empty($this->customValidators);
    }

    /**
     * Merge metadata into an export node array.
     */
    protected function addMetadataToNode(array &$node): void
    {
        if (!empty($this->metadata)) {
            $node['metadata'] = $this->metadata;
        }
    }

    /**
     * Get the PHP type name for a value, using AnyVali naming conventions.
     */
    protected static function getTypeName(mixed $value): string
    {
        if ($value === null) {
            return 'null';
        }
        if (is_bool($value)) {
            return 'boolean';
        }
        if (is_int($value)) {
            return 'number';
        }
        if (is_float($value)) {
            return 'number';
        }
        if (is_string($value)) {
            return 'string';
        }
        if (is_array($value)) {
            // Check if sequential (array) or associative (object)
            if (array_is_list($value)) {
                return 'array';
            }
            return 'object';
        }
        if (is_object($value)) {
            return 'object';
        }
        return gettype($value);
    }

    /**
     * Format a value for display in issue messages.
     */
    protected static function formatValue(mixed $value): string
    {
        if ($value === null) {
            return 'null';
        }
        if (is_bool($value)) {
            return $value ? 'true' : 'false';
        }
        if (is_int($value) || is_float($value)) {
            // For integers that are float due to JSON decode, show without decimal
            if (is_float($value) && floor($value) === $value && !is_infinite($value)) {
                return (string)(int)$value;
            }
            return (string)$value;
        }
        if (is_string($value)) {
            return $value;
        }
        return json_encode($value) ?: 'unknown';
    }
}
